compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIENormalizeLoopBounds.cpp

```diff
@@ -0,0 +1,191 @@
+      isZeroBased ? ub : rewriter.createOrFold<arith::SubIOp>(loc, ub, lb);
+  Value newUpperBound =
+      isStepOne ? diff
+                : rewriter.createOrFold<arith::CeilDivSIOp>(loc, diff, step);
+
+  Value newLowerBound = isZeroBased
+                            ? lb
+                            : rewriter.create<arith::ConstantOp>(
+                                  loc, rewriter.getZeroAttr(lb.getType()));
+  Value newStep = isStepOne
+                      ? step
+                      : rewriter.create<arith::ConstantOp>(
+                            loc, rewriter.getIntegerAttr(step.getType(), 1));
+
+  return {newLowerBound, newUpperBound, newStep};
+}
+
+/// Transform a `scf.for` loop with a strictly positive step
+///   for %i = %lb to %ub step %s
+/// into a 0-based loop with step 1
+///   for %ii = 0 to ceildiv(%ub - %lb, %s) step 1
+/// Insert an `affine.apply` operation to compute the denormalized index value.
+LogicalResult normalizeLoopBounds(RewriterBase &rewriter, scf::ForOp forOp) {
+  OpBuilder::InsertionGuard g(rewriter);
+  // Return if already normalized
+  std::optional<int64_t> lbInt = getConstantIntValue(forOp.getLowerBound());
+  std::optional<int64_t> stepInt = getConstantIntValue(forOp.getStep());
+  if (lbInt && stepInt && lbInt.value() == 0 && stepInt.value() == 1) {
+    return success();
+  }
+
+  Value iv = forOp.getInductionVar();
+  rewriter.setInsertionPoint(forOp);
+  auto newLoopParams =
+      emitNormalizedLoopBounds(rewriter, forOp.getLoc(), forOp.getLowerBound(),
+                               forOp.getUpperBound(), forOp.getStep());
+
+  rewriter.modifyOpInPlace(forOp, [&]() {
+    forOp.setLowerBound(newLoopParams.lowerBound);
+    forOp.setUpperBound(newLoopParams.upperBound);
+    forOp.setStep(newLoopParams.step);
+  });
+
+  rewriter.setInsertionPointToStart(forOp.getBody());
+  AffineExpr idx;
+  bindDims(forOp.getContext(), idx);
+  auto affineApply = rewriter.create<affine::AffineApplyOp>(
+      forOp.getLoc(), idx * stepInt.value() + lbInt.value(),
+      ValueRange{
+          iv,
+      });
+  SmallPtrSet<Operation *, 2> preserve(
+      {iv.getDefiningOp(), affineApply.getOperation()});
+  rewriter.replaceAllUsesExcept(iv, affineApply.getResult(), preserve);
+  return success();
+}
+
+/// Transform a `scf.forall` loop with a strictly positive steps
+///   forall (%i, %j) = (%lb0, %lb1) to (%ub0, %ub1) step (%s0, %s1)
+/// into a 0-based loop with step 1
+///   forall (%i, %j) in (ceildiv(%ub0 - %lb0, %s0), ceildiv(%ub1 - %lb1, %s1))
+/// Insert `affine.apply` operations to compute the denormalized index values.
+LogicalResult normalizeLoopBounds(RewriterBase &rewriter,
+                                  scf::ForallOp forallOp) {
+  OpBuilder::InsertionGuard g(rewriter);
+  if (forallOp.isNormalized()) return success();
+
+  SmallVector<OpFoldResult> newLbs;
+  SmallVector<OpFoldResult> newUbs;
+  SmallVector<OpFoldResult> newSteps;
+  for (auto &&[iv, lb, ub, step] : llvm::zip(
```

Comment:
The logic here should be reusable for `scf.for` by passing an `ArrayRef<OpFoldResult>` for the loop ranges. There is a `Range` struct provided here https://github.com/llvm/llvm-project/blob/89245b6cf808d4ed27cdea8d1e85fd9f4c6c844b/mlir/include/mlir/Dialect/Utils/StaticValueUtils.h#L33 that you can replace `LoopParams` with.